cms/djangoapps/contentstore/middleware.py

```diff
@@ -0,0 +1,77 @@
+"""
+Middlewares for contentstore app.
+"""
+
+import logging
+
+from django.core.exceptions import PermissionDenied
+from django.urls import Resolver404, resolve
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import AssetKey, CourseKey, UsageKey
+
+from common.djangoapps.student.roles import CourseLimitedStaffRole
+
+logger = logging.getLogger(__name__)
+
+
+class LimitedStaffAccessMiddleware:
+    """
+    Middleware to check access for limited staff users.
+    """
+
+    WHITELISTED_PAGES = ("/static",)
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    @staticmethod
+    def get_course_key(path):
+        """
+        Extracts and returns the course key from the given path.
+
+        Resolves the path and extract a key string, which can be a
+        course key, usage key, or asset key. Then tries to convert key string into
+        an appropriate key object (CourseKey, UsageKey, or AssetKey). If the key is either a usage
+        key or asset key, extracts and returns the associated course key. If the key is
+        invalid or the path cannot be resolved returns None.
+        """
+        try:
+            match = resolve(path)
+            kwargs = match.kwargs
+            key_string = (
+                kwargs.get("course_key_string")
+                or kwargs.get("usage_key_string")
+                or kwargs.get("course_id")
+                or kwargs.get("asset_key_string")
+                or kwargs.get("context_id")
+            )
+            if key_string:
+                for key_class in (CourseKey, UsageKey, AssetKey):
+                    try:
+                        key = key_class.from_string(key_string)
+                        return key.course_key if hasattr(key, "course_key") else key
+                    except InvalidKeyError:
+                        logger.warning("Invalid {}: {}.".format(key_class.__name__, key_string))
+        except Resolver404:
+            logger.warning("Path '{}' not resolved.".format(path))
+
+        return None
+
+    def __call__(self, request):
+        user = request.user
+        if user.is_authenticated and user.is_active:
+            path = request.path
+            if not any(
+                path.startswith(whitelisted_page)
+                for whitelisted_page in self.WHITELISTED_PAGES
+            ):
+                course_key = self.get_course_key(path)
+                if course_key and CourseLimitedStaffRole(course_key).has_user(user):
+                    logger.exception(
+                        "PermissionDenied: User '{}' attempted to access '{}' without sufficient permissions.".format(
+                            user, path
+                        )
+                    )
```

Comment:
Instead of checking for CourseLimitedStaffRole directly, we should probably be calling `get_user_permission(user, course_key)`, and then raise `PermissionDenied` if we get back `STUDIO_NO_PERMISSIONS`. That would cover `CourseLimitedStaffRole`, but also cover CCX courses, and possibly other outliers. And push us towards a more permission-centric stance than a "check the roles" stance.

Reply:
That was their first approach, but I suggested changing it to using the role instead because it didn't work for me. I'm re-testing in my environment to confirm there wasn't a misconfiguration on my side.